File: leetcode/easy/88-mergeSortedArray.java
class Solution {
    public void merge(int[] nums1, int m, int[] nums2, int n) {
        if (m == 1 && n == 1) {
            int temp1 = nums1[0];
            int temp2 = nums2[0];
            nums1[0] = Math.min(temp1, temp2);
            nums1[1] = Math.max(temp1, temp2);
            return;
        } 
            
        int i = m - 1;
        int j = n - 1;
        int k = nums1.length - 1;
        
        while (i >= 0 && j >= 0) {
            if (nums1[i] > nums2[j]) {
                nums1[k] = nums1[i];
                i--;
            } else {
                nums1[k] = nums2[j];
                j--;
            }
            k--;
        }
        
        while (i >= 0) {
            nums1[k] = nums1[i];
            i--; 
            k--;
        }
        
        while (j >= 0) {
            nums1[k] = nums2[j];
            j--;
            k--;
        } 
    }
}

File: leetcode/easy/88-mergeSortedArray.cpp
// first pass solution
class Solution {
public:
    void merge(vector<int>& nums1, int m, vector<int>& nums2, int n) {
        int i = m + n - 1; 
        int j = m - 1;
        int k = n - 1;

        while (i >= 0 && j >= 0 && k >= 0 ) {
            if (nums1[j] > nums2[k]) {
                nums1[i] = nums1[j];
                j--;
            } else {
                nums1[i] = nums2[k];
                k--;
            }
            i--;
        }

        while (i >= 0 && j >= 0) {
            nums1[i] = nums1[j];
            i--;
            j--;
        }

        while (i >= 0 && k >= 0) {
            nums1[i] = nums2[k];
            i--;
            k--;
        }
    }
};

// refactor
class Solution {
public:
    void merge(vector<int>& nums1, int m, vector<int>& nums2, int n) {
        int i = m + n - 1; 
        int j = m - 1;
        int k = n - 1;

        for (; i >= 0; i--) {
            if (j >= 0 && k >= 0) {
                if (nums1[j] > nums2[k]) {
                    nums1[i] = nums1[j];
                    j--;
                } else {
                    nums1[i] = nums2[k];
                    k--;
                }
            }

            else if (j >= 0) {
                nums1[i] = nums1[j];
                j--;
            }

            else {
                nums1[i] = nums2[k];
                k--;
            }
        }
    }
};
